### ISSUE TITLE
MultiViewer with Several captures per machine

### ISSUE THREAD

**Issue by yogch** [aol/moloch#459]
**Elasticsearch version**: 1.5.2

**Moloch version**: 0.11.5

**OS name and version**: Ubuntu 14.04

**Provide logs, stack traces and steps to reproduce**:

What happens when there're couple of machines, and on each machine there are couple of captures?
I want to see everything with MultiViewer.

Let's say I have 2 machines. On each machine 2 nodes:
machine 1: node1, node2
machine 2: node3, node4

From another machine, I want to expand the session detail of a session from node1 (call sessionDetail) and download PCAP.
The problem is, MultiViewer will try to connect to `node1`... instead of the IP of machine1.

Thanks

**Comment by awick** [comment]
https://github.com/aol/moloch/wiki/Architecture might help

You only need multies if you are running multiple moloch clusters, if you are only running 1 moloch clusters you do not need it.  A moloch cluster is defined by a separate ES cluster or the same ES cluster and different prefixes.

Assuming you are running 1 moloch cluster, just run multiple capture nodes per machine and 1 viewer per machine.

If you are really running mutliple moloch clusters and they are on the same machine, then you will need to set viewPort everywhere.

**Comment by yogch** [comment]
This is my architecture:

![image](https://cloud.githubusercontent.com/assets/13855845/14410650/e19d7128-ff3c-11e5-98eb-d272de575283.png)

On the machine with the MultiViewer, I want to download a pcap from the green cluster.
Specifically, from capture2

![image](https://cloud.githubusercontent.com/assets/13855845/14410679/49dde97a-ff3d-11e5-9b7f-26392ca98d7a.png)

The nodeName of capture2, is, `capture2`.

How does the MultiViewer know the `capture2` runs on this specific machine? when it should prepare the download link?

This is the call to MultiViewer:
![image](https://cloud.githubusercontent.com/assets/13855845/14410725/44319c1e-ff3e-11e5-994f-e7361f51ca8c.png)

You can see that this is node3. Then, the MultiViewer will try to issue a request to node3 (DNS).

This is not the machine name :rabbit:

**Comment by awick** [comment]
By default capture on node03 publishes it to ES.  Does "hostname" on the node03 machine produce something that multiviewer can lookup.

You can also override in the config file for multiviewer by setting viewerHost for node03

**Comment by yogch** [comment]
Did you try to use [elasticsearch's tribe node](https://www.elastic.co/guide/en/elasticsearch/reference/1.7/modules-tribe.html)? in order to do the aggregation of several elasticsearchs in a multi viewer environment?

I didn't see any `viewerHost` configuration.

**Comment by awick** [comment]
tribe node didn't exist when I created this.
tribe node has a limitation where the indexes have to be different names, which is now possible with prefixes, but wasn't possible before.

Maybe its viewHost.

So did you check if hostname returns a FQDN on all hosts?  That is really the easiest solution.

**Comment by yogch** [comment]
Do you mean the `gethostname()` function which settles the node name in `main.c` ?

According to my check it's equivalent to the `hostname` command, which brings the host name of the machine only, and not the full FQDN which can be retrieved by `hostname -f`...

**Comment by awick** [comment]
correct, hostname command and gethostname are the same.  On all our boxes hostname returns the fqdn, can you make it so it does that?  That will fix all your problems.

**Comment by yogch** [comment]
I need to check if it's possible here.

Anyway, it will fix part of the problem, no?
When I want to, lets say, download PCAP from the viewer.
The PCAP's name is settled by the node name, and date.
If I have a machine with two capture processes, how can the MultiViewer know which node inside the FQDN server to go to?
if `gethostname()` returns the name of the server, or the FQDN, it still doesn't differentiate between more than one node...

Am I correct?
Thanks

**Comment by awick** [comment]
I will look at how hard it is to fallback to the getdomainname call if gethostname isn't a FQDN.

There are lots of term confusion here.    Remember multiviewer is only needed when you are running MULTIPLE moloch clusters, NOT if you are running multiple capture processes on the same box.

On the machine with 2 moloch-capture nodes are they part of the same moloch cluster or not?

* If they are the same moloch cluster then
- multiviewer doesn't matter, and first you should get it working without using the multiviewer.  And if you are running just ONE moloch cluster you don't need multiviewer.
- there should only be ONE viewer.js running on the capture box
- it knows what file because it looks it up in the DB

* If the two moloch-capture processes are part of TWO DIFFERENT moloch clusters then 
- you do need multiviewer.
- you will need to run TWO viewer.js on the box, one for each cluster, best to do with 2 config files one for each cluster
- you should set a viewPort= on both config files
- that same config file needs to be on the multiviewer

**Comment by yogch** [comment]
Ok,

I think I got it...
I thought the viewer.js knows the hostname according to the node name.
I just realized that it knows the hostname using another index, called `stats`, which maps between the node name and the host name.
Is it correct?

So, the capture who writes the `stats`, writes the mapping between node name and hostname,
and then the MultiViewer would be able to connect to the node, according to the hostname written in the stats.

So the `gethostname()`, which settles `config.hostName`, should return FQDN.

Thanks

**Comment by awick** [comment]
correct

**Comment by yogch** [comment]
Great! :dancer:

**Comment by yogch** [comment]
> tribe node has a limitation where the indexes have to be different names, which is now possible with prefixes

How is it possible now with prefix? I couldn't make it work.

**Comment by awick** [comment]
How is what possible?  Are you asking if tribe nodes work with Moloch?  Moloch doesn't work with tribe nodes.  You asked why I didn't use tribe nodes and I was explaining at the time I created Moloch that tribe nodes didn't exist and even when tribe nodes came around it wasn't to use because tribe nodes required different index names, which wasn't even possible with moloch until recently when prefixes were added.  However I still haven't made moloch support tribe nodes.
